File: frontend/src/api/diagram.ts
import type { DiagramRequest, DiagramResponse } from '@/types/diagram'
import { apiClient, getApiErrorMessage } from '@/lib/axios'

export async function generateRepositoryDiagram(
  repositoryId: number,
  payload: DiagramRequest,
) {
  try {
    const { data } = await apiClient.post<DiagramResponse>(
      `/api/v1/repositories/${repositoryId}/diagram`,
      payload,
    )
    return data
  } catch (error) {
    throw new Error(getApiErrorMessage(error, 'Diagram generation failed'))
  }
}

File: backend/api/routes/diagram.py
"""Repository Mermaid diagram API."""

from fastapi import APIRouter

from diagrams.service import generate_repository_diagram
from schemas.diagram import DiagramRequest, DiagramResponse

router = APIRouter(prefix="/repositories/{repository_id}/diagram", tags=["diagram"])


@router.post("", response_model=DiagramResponse)
def create_repository_diagram(repository_id: int, payload: DiagramRequest) -> DiagramResponse:
    """Generate a Mermaid import flowchart from the knowledge graph (LLM fallback if needed)."""
    result = generate_repository_diagram(
        repository_id,
        message=payload.message,
        limit=payload.limit,
    )
    return DiagramResponse.model_validate(result)
